File: xabber/src/main/java/com/xabber/android/data/extension/groupchat/CreateGroupchatIQ.java
package com.xabber.android.data.extension.groupchat;

import com.xabber.android.data.entity.AccountJid;
import com.xabber.android.data.entity.ContactJid;
import com.xabber.android.data.message.chat.groupchat.GroupchatIndexType;
import com.xabber.android.data.message.chat.groupchat.GroupchatManager;
import com.xabber.android.data.message.chat.groupchat.GroupchatMembershipType;
import com.xabber.android.data.message.chat.groupchat.GroupchatPrivacyType;
import com.xabber.xmpp.SimpleNamedElement;

import org.jivesoftware.smack.packet.IQ;
import org.jivesoftware.smack.packet.NamedElement;
import org.jivesoftware.smack.util.XmlStringBuilder;
import org.jxmpp.jid.Jid;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class CreateGroupchatIQ extends IQ {

    public static final String HASH_BLOCK = "#create";
    public static final String NAMESPACE = GroupchatManager.NAMESPACE.concat(HASH_BLOCK);
    public static final String QUERY_ELEMENT = "query";
    public static final String JID_ELEMENT = "jid";
    private static final String NAME_ELEMENT = "name";
    private static final String DESCRIPTION_ELEMENT = "description";
    private static final String LOCALPART_ELEMENT = "localpart";
    private static final String PRIVACY_ELEMENT = "privacy";
    private static final String INDEX_ELEMENT = "index";
    private static final String MEMBERSHIP_ELEMENT = "membership";
    private static final String CONTACTS_ELEMENT = "contacts";
    private static final String CONTACT_ELEMENT = "contact";

    private String to;
    private Jid from;
    private Collection<NamedElement> elements = new ArrayList<>();

    @Override
    public Jid getFrom() {
        return from;
    }

    @Override
    public Type getType() {
        return Type.set;
    }

    public CreateGroupchatIQ(Jid from, String to, String groupName, String groupLocalpart,
                             String description, GroupchatMembershipType membershipType,
                             GroupchatPrivacyType privacyType, GroupchatIndexType indexType){
        super(QUERY_ELEMENT, NAMESPACE);
        this.setType(Type.set);
        this.setFrom(from);
        this.setTo(to);

        elements.add(new SimpleNamedElement(NAME_ELEMENT, groupName));
        elements.add(new SimpleNamedElement(DESCRIPTION_ELEMENT, description));
        elements.add(new SimpleNamedElement(MEMBERSHIP_ELEMENT, membershipType.toXml()));
        elements.add(new SimpleNamedElement(PRIVACY_ELEMENT, privacyType.toXml()));
        elements.add(new SimpleNamedElement(INDEX_ELEMENT, indexType.toXml()));
        elements.add(new MSimpleParentElement(CONTACTS_ELEMENT,
                new SimpleNamedElement(CONTACT_ELEMENT, from.asBareJid().toString())));
        if (groupLocalpart != null && !groupLocalpart.isEmpty())
            elements.add(new SimpleNamedElement(LOCALPART_ELEMENT, groupLocalpart));

        this.from = from;
        this.to = to;
    }

    @Override
    protected IQChildElementXmlStringBuilder getIQChildElementBuilder(IQChildElementXmlStringBuilder xml) {
        xml.rightAngleBracket();
        if (elements.size() != 0)
            for (NamedElement element : elements)
                xml.append(element.toXML());
        return xml;
    }

    private static class MSimpleParentElement implements NamedElement{
        String elementName;
        List<NamedElement> nestedElements = new ArrayList<>();

        MSimpleParentElement(String elementName, List<NamedElement> nestedElements){
            this.elementName = elementName;
            this.nestedElements.addAll(nestedElements);
        }

        MSimpleParentElement(String elementName, NamedElement nestedElement){
            this.elementName = elementName;
            this.nestedElements.add(nestedElement);
        }

        @Override
        public String getElementName() { return elementName; }

        @Override
        public CharSequence toXML() {
            XmlStringBuilder xmlStringBuilder = new XmlStringBuilder(this);
            xmlStringBuilder.rightAngleBracket();
            for (NamedElement nestedElement : nestedElements)
                xmlStringBuilder.append(nestedElement.toXML());
            xmlStringBuilder.closeElement(this);
            return xmlStringBuilder;
        }
    }

    public interface CreateGroupchatIqResultListener{
        void onJidConflict();
        void onOtherError();
        void onSuccessfullyCreated(AccountJid accountJid, ContactJid contactJid);
    }

}

File: xabber/src/main/java/com/xabber/android/data/extension/groupchat/create/CreateGroupchatIQ.kt
package com.xabber.android.data.extension.groupchat.create

import com.xabber.android.data.message.chat.groupchat.GroupchatIndexType
import com.xabber.android.data.message.chat.groupchat.GroupchatMembershipType
import com.xabber.android.data.message.chat.groupchat.GroupchatPrivacyType
import com.xabber.xmpp.SimpleNamedElement
import org.jivesoftware.smack.packet.IQ
import org.jivesoftware.smack.packet.NamedElement
import org.jivesoftware.smack.util.XmlStringBuilder
import org.jxmpp.jid.Jid

class CreateGroupchatIQ(val contact: Jid, to: String, val groupName: String?, private val groupLocalpart: String?,
                        val description: String?, private val membershipType: GroupchatMembershipType,
                        private val privacyType: GroupchatPrivacyType, private val indexType: GroupchatIndexType)
    : GroupchatCreateAbstractIQ() {

    override fun getIQChildElementBuilder(xml: IQChildElementXmlStringBuilder) = xml.apply {
        rightAngleBracket()

        append(SimpleNamedElement(NAME_ELEMENT, groupName!!).toXML())
        append(SimpleNamedElement(DESCRIPTION_ELEMENT, description!!).toXML())
        append(SimpleNamedElement(MEMBERSHIP_ELEMENT, membershipType.toXml()!!).toXML())
        append(SimpleNamedElement(PRIVACY_ELEMENT, privacyType.toXml()!!).toXML())
        append(SimpleNamedElement(INDEX_ELEMENT, indexType.toXml()!!).toXML())

        append(ContactsNamedElement(contact.toString()).toXML())

        if (groupLocalpart != null && groupLocalpart.isNotEmpty())
            append(SimpleNamedElement(LOCALPART_ELEMENT, groupLocalpart).toXML())

    }

    private class ContactsNamedElement(val contactJid: String) : NamedElement {

        companion object {
            private const val CONTACTS_ELEMENT = "contacts"
            private const val CONTACT_ELEMENT = "contact"
        }

        override fun getElementName() = CONTACTS_ELEMENT

        override fun toXML() = XmlStringBuilder(this).apply {
            rightAngleBracket()
            append(SimpleNamedElement(CONTACT_ELEMENT, contactJid).toXML())
            closeElement(CONTACTS_ELEMENT)
        }
    }

    companion object {
        const val HASH_BLOCK = "#create"
        const val ELEMENT_LOCALPART = "localpart"
        private const val NAME_ELEMENT = "name"
        private const val DESCRIPTION_ELEMENT = "description"
        private const val LOCALPART_ELEMENT = "localpart"
        private const val PRIVACY_ELEMENT = "privacy"
        private const val INDEX_ELEMENT = "index"
        private const val MEMBERSHIP_ELEMENT = "membership"
    }

    init {
        this.type = Type.set
        this.from = contact
        this.setTo(to)
    }

    /**
     * WARN! This is not a fully implemented class
     */
    class ResultIq(val localpart: String) : IQ(ELEMENT, NAMESPACE + HASH_BLOCK){
        val jid: String
            get() = "$localpart@$from"

        override fun getIQChildElementBuilder(xml: IQChildElementXmlStringBuilder?) = xml
    }

}